### DISCUSSION TITLE
Unsolicited on_cancel

### DISCUSSION THREAD

**Discussion by venkatramanm** [beckn/protocol-specifications#276] [Core Specification]
A Situation:
* Seller cancels an order for lack of inventory or any other reason 
* BPP calls BAP's on_cancel. !
* BAP Sends ACK 
* But async  processing of BAP fails as BPP didnot honor some agreement 
* BAP Thinks order will be fulfilled, BPP thinks order has been cancelled. 

-- Stale mate-- 
* Dual signed is needed for cancel just like for confirm. 

Options:
1. May be BAP should process cancel updates on its side synchronously before sending ACK. !?
2. May be BAP should call "on_on_cancel" on BPP to close loop. Which finishes the dual signing of cancel contract.

**Comment by venkatramanm** [comment]
May be all "on_.xyz" calls must be done synchronously at BAP end. 
Another situation:
1. BAP calls "confirm"
2. BPP sends ACK 
3. BPP processes confirmation at its end  async
4. BPP makes on_confirm callback 
5. BAP sends ACK 
6. BAP's async process fails. 
7. BAP Thinks BPP has not yer confirmed. BPP thinks order has been nicely confirmed. 

May be we should document that "on_xyz" calls at BAP must be done synchronously before sending "ACK" 
Basically if async processing of BAP's API fails, There is a miscommunication.
